lldb-eval/eval_test.cc

```diff
@@ -492,6 +484,55 @@ TEST_F(EvalTest, TestPointerArithmetic) {
                       "'const char *')"));
 }
 
+TEST_F(EvalTest, PointerPointerComparison) {
+  EXPECT_THAT(Eval("p_void == p_void"), IsEqual("true"));
+  EXPECT_THAT(Eval("p_void == p_char1"), IsEqual("true"));
+  EXPECT_THAT(Eval("p_void != p_char1"), IsEqual("false"));
+  EXPECT_THAT(Eval("p_void > p_char1"), IsEqual("false"));
+  EXPECT_THAT(Eval("p_void >= p_char1"), IsEqual("true"));
+  EXPECT_THAT(Eval("p_void < (p_char1 + 1)"), IsEqual("true"));
+  EXPECT_THAT(Eval("pp_void0 + 1 == pp_void1"), IsEqual("true"));
+
+  EXPECT_THAT(Eval("(void*)1 == (void*)1"), IsEqual("true"));
+  EXPECT_THAT(Eval("(void*)1 != (void*)1"), IsEqual("false"));
+  EXPECT_THAT(Eval("(void*)2 > (void*)1"), IsEqual("true"));
+  EXPECT_THAT(Eval("(void*)2 < (void*)1"), IsEqual("false"));
+}
+
+TEST_F(EvalTest, PointerIntegerComparison) {
+  EXPECT_THAT(Eval("(void*)0 == 0"), IsEqual("true"));
+  EXPECT_THAT(Eval("0 != (void*)0"), IsEqual("false"));
+
+  EXPECT_THAT(Eval("(void*)0 == nullptr"), IsEqual("true"));
+  EXPECT_THAT(Eval("(void*)0 != nullptr"), IsEqual("false"));
+  EXPECT_THAT(Eval("nullptr == (void*)1"), IsEqual("false"));
+  EXPECT_THAT(Eval("nullptr != (void*)1"), IsEqual("true"));
+
+  EXPECT_THAT(Eval("nullptr == nullptr"), IsEqual("true"));
+  EXPECT_THAT(Eval("nullptr != nullptr"), IsEqual("false"));
+  EXPECT_THAT(Eval("nullptr == 0"), IsEqual("true"));
+  EXPECT_THAT(Eval("0 != nullptr"), IsEqual("false"));
+
+  EXPECT_THAT(
+      Eval("(void*)0 > nullptr"),
+      IsError(
+          "invalid operands to binary expression ('void *' and 'nullptr_t')"));
+
+  // These are not allowed by C++, but we support it as an extension.
+  EXPECT_THAT(Eval("(void*)1 == 1"), IsEqual("true"));
+  EXPECT_THAT(Eval("(void*)1 == 0"), IsEqual("false"));
+  EXPECT_THAT(Eval("(void*)1 > 0"), IsEqual("true"));
+  EXPECT_THAT(Eval("(void*)1 < 0"), IsEqual("false"));
+  EXPECT_THAT(Eval("1 > (void*)0"), IsEqual("true"));
+  EXPECT_THAT(Eval("2 < (void*)3"), IsEqual("true"));
+
+  // Integer is converted to uintptr_t, so negative numbers because large
+  // positive numbers.
+  EXPECT_THAT(Eval("(void*)0xffffffffffffffff == -1"), IsEqual("true"));
```

Comment:
```suggestion
  EXPECT_THAT(Eval("(void*)0xffff'ffff'ffff'ffff == -1"), IsEqual("true"));
```

I have a feeling this might not be 100% portable. It certainly works on 32-bit and 64-bit machines, but what if 128-bit machines become a thing? (Then again, this might take a long time 😛).